daffodil-runtime2/src/main/scala/org/apache/daffodil/runtime2/generators/CodeGeneratorState.scala

```diff
@@ -404,76 +542,117 @@ class CodeGeneratorState(private val root: ElementBase) {
+    if (hasArray) pushArray(child)
 
+    // Generate statements to initialize, parse, and unparse the child
+    Runtime2CodeGenerator.generateCode(child.enclosedElement, this)
+    if (!child.isSimpleType) {
       if (hasChoice)
-        structs.top.initChoiceStatements += initChoiceStatement
+        structs.top.initChoiceStatements += s"$indent2        ${C}_initERD(&instance->$e$deref, (InfosetBase *)instance);"
       else
-        structs.top.initERDStatements += initERDStatement
-      structs.top.initSelfStatements += initSelfStatement
-      structs.top.parserStatements += parseStatement
-      structs.top.unparserStatements += unparseStatement
+        structs.top.initERDStatements += s"$indent2    ${C}_initERD(&instance->$e$deref, (InfosetBase *)instance);"
+      structs.top.parserStatements +=
+        s"""$indent1$indent2    ${C}_parseSelf(&instance->$e$deref, pstate);
+           |$indent1$indent2    if (pstate->error) return;""".stripMargin
+      structs.top.unparserStatements +=
+        s"""$indent1$indent2    ${C}_unparseSelf(&instance->$e$deref, ustate);
+           |$indent1$indent2    if (ustate->error) return;""".stripMargin
     }
-    if (arraySize > 0)
-      for (i <- 0 until arraySize)
-        addStatements(s"[$i]")
-    else
-      addStatements("")
 
-    if (hasChoice) {
-      val initChoiceStatement = s"        break;"
-      val parseStatement = s"        break;"
-      val unparseStatement = s"        break;"
+    // Generate statements to align the next element
+    val mustAlign = child.endingAlignmentApprox.nBits > child.priorAlignmentApprox.nBits
+    if (mustAlign) {
+      val nBits = child.endingAlignmentApprox.nBits
+      structs.top.parserStatements +=
+        s"""$indent1$indent2    // Fill to closest alignment
+           |$indent1$indent2    parse_fill_bits((pstate->bitPos0b + ${nBits - 1}) / $nBits * $nBits, pstate);
+           |$indent1$indent2    if (pstate->error) return;""".stripMargin
+      structs.top.unparserStatements +=
+        s"""$indent1$indent2    // Fill to closest alignment
+           |$indent1$indent2    unparse_fill_bits((ustate->bitPos0b + ${nBits - 1}) / $nBits * $nBits, '\\0', ustate);
+           |$indent1$indent2    if (ustate->error) return;""".stripMargin
+    }
 
-      structs.top.initChoiceStatements += initChoiceStatement
-      structs.top.parserStatements += parseStatement
-      structs.top.unparserStatements += unparseStatement
+    // Terminate the array if the child occurs multiple times
+    if (hasArray) popArray(child)
+
+    // Generate a choice statement break if the child is in a choice element
+    if (hasChoice) {
+      val break = s"        break;"
+      structs.top.initChoiceStatements += break
+      structs.top.parserStatements += break
+      structs.top.unparserStatements += break
     }
   }
 
+  private def pushArray(e: ElementBase): Unit = {
+    val C = structs.top.C
+    val name = "%5Bi%5D"
```

Comment:
This is holdover code from an earlier attempt to convert a DPath occursCount expression to C code.  Without going into too many details, I tried to use the name field to build a normalized URL before converting it to C struct dot notation, but that approach needed to use an `[i]` dereference with an i of the right value which wasn't feasible.  I'll remove the `name` field since I was keeping it only for debugging to help see the difference between duplicate "structs" with the same class (C) in the stack at this point.